File: test/ballot_proof_poseidon_test.go
package test

import (
	"encoding/json"
	"fmt"
	"log"
	"math"
	"math/big"
	"os"
	"testing"

	"github.com/vocdoni/z-ircuits/utils"
	"go.vocdoni.io/dvote/crypto/ethereum"
	"go.vocdoni.io/dvote/util"
)

func TestBallotProofPoseidon(t *testing.T) {
	if persist && testID == "" {
		t.Error("Test ID is required when persisting")
		return
	}

	acc := ethereum.NewSignKeys()
	if err := acc.Generate(); err != nil {
		t.Error(err)
		return
	}
	var (
		address   = acc.Address().Bytes()
		processID = util.RandomBytes(20)
		// ballot inputs
		fields          = utils.GenerateBallotFields(5, 16, 0, false)
		n_fields        = 8
		maxCount        = 5
		forceUniqueness = 0
		maxValue        = 16
		minValue        = 0
		costExp         = 2
		costFromWeight  = 0
		weight          = 0
		// circuit assets
		wasmFile = "../artifacts/ballot_proof_poseidon_test.wasm"
		zkeyFile = "../artifacts/ballot_proof_poseidon_test_pkey.zkey"
		vkeyFile = "../artifacts/ballot_proof_poseidon_test_vkey.json"
	)
	// encrypt ballot
	_, pubKey := utils.GenerateKeyPair()
	k, err := utils.RandomK()
	if err != nil {
		t.Errorf("Error generating random k: %v\n", err)
		return
	}
	// generate vote ID
	bigPID := util.BigToFF(new(big.Int).SetBytes(processID))
	bigAddr := util.BigToFF(new(big.Int).SetBytes(address))
	voteID, err := utils.VoteID(bigPID, bigAddr, k)
	if err != nil {
		t.Errorf("Error generating vote ID: %v\n", err)
		return
	}
	// encrypt ballot fields and get them in plain format
	cipherfields, plainCipherfields := utils.CipherBallotFields(fields, n_fields, pubKey, k)
	bigInputs := []*big.Int{
		bigPID,
		big.NewInt(int64(maxCount)),
		big.NewInt(int64(forceUniqueness)),
		big.NewInt(int64(maxValue)),
		big.NewInt(int64(minValue)),
		big.NewInt(int64(math.Pow(float64(maxValue), float64(costExp))) * int64(maxCount)),
		big.NewInt(int64(maxCount)),
		big.NewInt(int64(costExp)),
		big.NewInt(int64(costFromWeight)),
		pubKey.X,
		pubKey.Y,
		bigAddr,
		voteID,
	}
	bigInputs = append(bigInputs, plainCipherfields...)
	bigInputs = append(bigInputs, big.NewInt(int64(weight)))
	inputsHash, err := utils.MultiPoseidon(bigInputs...)
	if err != nil {
		log.Fatalf("Error hashing: %v\n", err)
		return
	}
	// circuit inputs
	inputs := map[string]any{
		"fields":            utils.BigIntArrayToStringArray(fields, n_fields),
		"num_fields":        fmt.Sprint(maxCount),
		"unique_values":     fmt.Sprint(forceUniqueness),
		"max_value":         fmt.Sprint(maxValue),
		"min_value":         fmt.Sprint(minValue),
		"max_value_sum":     fmt.Sprint(int(math.Pow(float64(maxValue), float64(costExp))) * maxCount), // (maxValue)^costExp * maxCount
		"min_value_sum":     fmt.Sprint(maxCount),
		"cost_exponent":     fmt.Sprint(costExp),
		"cost_from_weight":  fmt.Sprint(costFromWeight),
		"address":           bigAddr.String(),
		"weight":            fmt.Sprint(weight),
		"process_id":        bigPID.String(),
		"vote_id":           voteID.String(),
		"encryption_pubkey": []string{pubKey.X.String(), pubKey.Y.String()},
		"k":                 k.String(),
		"cipherfields":      cipherfields,
		"inputs_hash":       inputsHash.String(),
	}
	bInputs, _ := json.MarshalIndent(inputs, "  ", "  ")
	t.Log("Inputs:", string(bInputs))
	proofData, pubSignals, err := utils.CompileAndGenerateProof(bInputs, wasmFile, zkeyFile)
	if err != nil {
		t.Errorf("Error compiling and generating proof: %v\n", err)
		return
	}
	t.Log("Proof:", proofData)
	t.Log("Public signals:", pubSignals)
	// read zkey file
	vkey, err := os.ReadFile(vkeyFile)
	if err != nil {
		t.Errorf("Error reading zkey file: %v\n", err)
		return
	}
	if err := utils.VerifyProof(proofData, pubSignals, vkey); err != nil {
		t.Errorf("Error verifying proof: %v\n", err)
		return
	}
	log.Println("Proof verified")
	if persist {
		// try to create the directory if it doesn't exist
		if _, err := os.Stat(path); os.IsNotExist(err) {
			if err := os.Mkdir(path, 0o755); err != nil {
				t.Errorf("Error creating directory: %v\n", err)
				return
			}
		}
		if err := os.WriteFile(fmt.Sprintf("%s/%s_proof.json", path, testID), []byte(proofData), 0o644); err != nil {
			t.Errorf("Error writing proof file: %v\n", err)
			return
		}
		if err := os.WriteFile(fmt.Sprintf("%s/%s_pub_signals.json", path, testID), []byte(pubSignals), 0o644); err != nil {
			t.Errorf("Error writing public signals file: %v\n", err)
			return
		}
	}
}

File: test/ts/ballot_proof_poseidon_test.ts
import { hexToField, multiHash, encrypt, prove, verify, voteId } from './utils';

(async () => {
    const wasm = "../../artifacts/ballot_proof_poseidon_test.wasm";
    const pk = "../../artifacts/ballot_proof_poseidon_test_pkey.zkey";
    const vk = "../../artifacts/ballot_proof_poseidon_test_vkey.json";
    // encrypt fields
    const pubKey: [bigint, bigint] = [
        BigInt("14683031697277856265190472023105198820413415816394525437006041413571428119286"),
        BigInt("8231930431069649913978957226360235712312621187451969137400305691913512440493")
    ];
    const k: bigint = BigInt("650538809577380042220943083323036196153738464670981167525900386056100355632");
    const n_fields = 8;
    const fields = [1, 2, 3, 4, 5];
    // fill with zeros to reach n_fields both fields and cipherfields
    const cipherfields: BigInt[][][] = new Array(n_fields).fill(0).map(() => [
        [BigInt(0), BigInt(0)],
        [BigInt(0), BigInt(0)],
    ]);
    for (let i = 0; i < n_fields; i++) {
        if (i < fields.length) {
            cipherfields[i] = encrypt(BigInt(fields[i]), pubKey, k)
        } else {
            fields.push(0);
        }
    }
    // compute nullifier
    const address = hexToField("0x6Db989fbe7b1308cc59A27f021e2E3de9422CF0A");
    const process_id = hexToField("0xf16236a51F11c0Bf97180eB16694e3A345E42506");
    const vote_id = await voteId(process_id, address, k);
    // init inputs
    const inputs = {
        fields,
        num_fields: 5,
        unique_values: 1,
        max_value: 5 + 1,
        min_value: 0,
        max_value_sum: 56,
        min_value_sum: 5,
        cost_exponent: 2,
        cost_from_weight: 0,
        process_id,
        vote_id,
        weight: 0,
        encryption_pubkey: pubKey,
        k,
        cipherfields,
        address,
        inputs_hash: BigInt(0)
    };
    const bigFields: bigint[] = [];
    for (let i = 0; i < fields.length; i++) {
        bigFields.push(BigInt(fields[i]));
    }
    const plainBigCipherFields: bigint[] = [];
    for (let i = 0; i < cipherfields.length; i++) {
        plainBigCipherFields.push(cipherfields[i][0][0] as bigint);
        plainBigCipherFields.push(cipherfields[i][0][1] as bigint);
        plainBigCipherFields.push(cipherfields[i][1][0] as bigint);
        plainBigCipherFields.push(cipherfields[i][1][1] as bigint);
    }
    inputs.inputs_hash = multiHash([
        inputs.process_id,
        BigInt(inputs.num_fields),
        BigInt(inputs.unique_values),
        BigInt(inputs.max_value),
        BigInt(inputs.min_value),
        BigInt(inputs.max_value_sum),
        BigInt(inputs.min_value_sum),
        BigInt(inputs.cost_exponent),
        BigInt(inputs.cost_from_weight),
        inputs.encryption_pubkey[0],
        inputs.encryption_pubkey[1],
        address,
        inputs.vote_id,
        ...plainBigCipherFields,
        BigInt(inputs.weight),
    ]);
    console.log("inputs", inputs);
    // generate proof
    const [proof, publicSignals] = await prove(inputs, wasm, pk);
    console.log("proof", proof);
    console.log("pubSignals", publicSignals);
    // verify proof
    const verified = await verify(vk, proof, publicSignals);
    console.log("Proof verified?", verified);
    // exit
    process.exit();
})();